### PR TITLE
[STRATCONN-14] DC Floodlight Pass Top Level UserId on Conversion Event

### PR DESCRIPTION
**Pull request by briemcnally**
JIRA: https://segment.atlassian.net/browse/STRATCONN-14
JIRA: https://segment.atlassian.net/browse/STRATCONN-12
JIRA: https://segment.atlassian.net/browse/STRATCONN-86
CC: https://segment.atlassian.net/browse/CC-6331

**What does this PR do?**
**STRATCONN-14**
Currently users are reporting that when they fire an `identify()` event for a user, on subsequent `track()` calls that top level `userId()` is not being sent to DC Floodlight, only the `userId()` nested in properties within the `track()` call itself is being sent. 

This PR adds functionality specifically for `userId`, where if the segmentProp from settings is `{{userId}}` indicating the customer wants to use the top level `userId()` (which would only be present if the customer fired an identify event on the window), then check `self.analytics.user().id()`, otherwise proceed with existing functionality. 

**NOTE** this change is specific for calls where an `identify()` call was previously made and the `userId()` is stored on the analytics object **AND** the customer wants to map the top level `userId()` field as seen from their Segment debugger (which is the `userId()` on the analytics object). 

**STRATCONN-12**
Adds support for products array Custom Variables. Customer confirmed that DCF would expect the products array to pass through as a comma separated array. For example, say you whitelisted products.$.sku  to the custom Floodlight variable u1​ and passed a call like the one below:
```
analytics.track('Checkout Started', {
sku: '1324abcd',
products: [ { sku: 'sku1' } , { sku: 'sku2' } , { sku: 'sku3' } ]
});
```
DCF would expect the the u1​ parameter to be set as a comma-separated array (example:  `[sku1,sku2,sku3]`​).

**STRATCONN-86**
DCF supports two methods of counting: Items Sold and Transaction based. Currently our DCF Integration only supports Items Sold. Current implementation sets tagParams.qty by # of products in products array, falling back to checking for an actual property called quantity.

For this new change we will implement a new setting useTransactionCounting that will default to false as to not cause any regressions for current customers. When this setting is enabled, and the event is a Sales Tag, we will always default to set to tagParams.qty to 1 per the Transaction counting method.

**Are there breaking changes in this PR?**
No, this was you cans see changes validated below: 
Screen Shot 2020-04-16 at 12 01 14 PM

**Any background context you want to provide?**
N/A

**Is there parity with the server-side/android/iOS integration components (if applicable)?**
N/A

**Does this require a new integration setting? If so, please explain how the new setting works**
No

**Links to helpful docs and other external resources**
N/A

### CODE REVIEW

File: integrations/doubleclick-floodlight/lib/index.js [hunk 1]
```diff
@@ -105,7 +106,11 @@ Floodlight.prototype.track = function(track) {
 
         if (Array.isArray(segmentProp)) {
           segmentProp = segmentProp.pop();
-          segmentPropValue = find(track.json(), segmentProp);
+          if (segmentProp === 'userId') {
```
Comment (gpsamson) [inline integrations/doubleclick-floodlight/lib/index.js:109]: Is there any reason we shouldn't support other top-level values via `track.proxy()`?

Comment (briemcnally) [reply integrations/doubleclick-floodlight/lib/index.js:109]: They are supported by this: `segmentPropValue = find(track.json(), segmentProp);` which is the existing behavior. It is just that `userId` is stored on the analytics object not the track.

Comment (gpsamson) [reply integrations/doubleclick-floodlight/lib/index.js:109]: Ahh, okay. Got it.

Comment (briemcnally) [reply integrations/doubleclick-floodlight/lib/index.js:109]: Updated the description of the change change to make it a little more explicit/clear. 👍

**Comment by briemcnally** [comment]
@gpsamson I added support for nested products variables. Ie. when a customer indicated `products.$.price` to be mapped to `u8`. The feature request was that these be sent as a comma separated array. https://segment.atlassian.net/browse/STRATCONN-12

File: integrations/doubleclick-floodlight/lib/index.js [hunk 2]
```diff
@@ -101,11 +102,27 @@ Floodlight.prototype.track = function(track) {
...
+        var segmentPropValue;
+        if (Array.isArray(segmentProp) && segmentProp[0] === 'products') {
+          segmentProp = segmentProp.pop();
+          var productPropArray = [];
+          each(function(product) {
+            if (product[segmentProp]) {
+              productPropArray.push(product[segmentProp]);
+            }
+          }, track.products());
+          segmentPropValue = productPropArray;
```
Comment (gpsamson) [inline integrations/doubleclick-floodlight/lib/index.js:118]: Does this need to be joined?

Comment (briemcnally) [reply integrations/doubleclick-floodlight/lib/index.js:118]: In Stratconn-12 the customer requested that it be sent as a comma separated array and not a string: "DCF would expect the the u1​ parameter to be set as a comma-separated array (example:  [sku1,sku2,sku3]​)." DCF docs are pretty lacking on this topic but happy to follow up with the CSM on expected behavior before deploying.

Comment (gpsamson) [reply integrations/doubleclick-floodlight/lib/index.js:118]: Got it. Sorry, I misread your comment above.

**Pull request merged**